In [1]:
import pandas as pd
import numpy as np

data=pd.read_excel('D:\coding\PRML\hobby\cervical cancer\lugol_glcm_properties.xlsx')
data.head()

Image   contrast  dissimilarity  homogeneity    energy  correlation  \
0  100.jpg  68.391943       3.442358     0.407688  0.086368     0.985854   
1  101.jpg  43.106726       2.694521     0.447988  0.072553     0.987294   
2  102.jpg  54.911871       3.011028     0.445394  0.083001     0.986244   
3  103.jpg  42.799076       3.005844     0.440684  0.094325     0.990842   
4  104.jpg  50.750850       4.431428     0.233907  0.024835     0.969365   

    entropy  
0  6.901329  
1  6.636268  
2  6.722089  
3  6.823557  
4  6.796737

In [6]:
df.drop(columns=["CaseNumber","SCJ","SCJ Location","Acitowhite area","Acitowhite area color","Acitowhite area margin","Acitowhite area surface","Acitowhite area location","Acitowhite area size","Eligibility for ablative treatment","Histology findings"],inplace=True)
df.drop(index=185,inplace=True)
df

contrast  dissimilarity  homogeneity    energy  correlation   entropy  \
0    283.744565      11.801893     0.165723  0.059224     0.972048  7.741566   
1    155.903369       6.292197     0.320660  0.179214     0.986113  7.093800   
2     78.084047       4.093255     0.350709  0.064911     0.988939  7.466649   
3     70.028251       4.444391     0.350128  0.074385     0.994013  7.624819   
4     65.833814       4.335129     0.324166  0.043489     0.994367  7.794829   
..          ...            ...          ...       ...          ...       ...   
180   28.602161       2.794230     0.376036  0.040077     0.994485  7.432741   
181   24.775397       2.788929     0.365146  0.023337     0.995700  7.703034   
182   34.473653       3.404823     0.328069  0.024295     0.995153  7.774326   
183   29.405277       2.902007     0.385178  0.057611     0.996037  7.478304   
184   50.114054       3.638373     0.363525  0.057077     0.994712  7.643260   

    CaseID                   VIA  
0      AFC              Negative  
1      AJL              Negative  
2      AGY              Negative  
3      AJE              Negative  
4      AHS              Negative  
..     ...                   ...  
180    AAC  Suspicious of cancer  
181    AAR  Suspicious of cancer  
182    AEY  Suspicious of cancer  
183    AEU  Suspicious of cancer  
184    ADF  Suspicious of cancer  

[185 rows x 8 columns]

In [7]:
from sklearn.model_selection import train_test_split

X=df.drop(columns=["CaseID","VIA"])
Y=df["VIA"]
x_train,x_test,y_train,y_test=train_test_split(X,Y,test_size=0.2,random_state=42)
x_train.head()

contrast  dissimilarity  homogeneity    energy  correlation   entropy
65   72.795115       4.598985     0.317680  0.082471     0.991115  7.441295
67   67.017911       3.712494     0.419994  0.036050     0.993991  7.757485
31  112.194133       5.904033     0.296427  0.094097     0.990220  7.605158
12   79.972613       4.532531     0.305087  0.035221     0.993216  7.834926
41   36.861079       2.949612     0.381122  0.027213     0.992183  7.441585

In [8]:
from sklearn.svm import SVC 
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
svm = SVC(kernel = 'linear', C = 1).fit(x_train, y_train) 
svm_predictions = svm.predict(x_test) 
  
# model accuracy for X_test   
accuracy = svm.score(x_test, y_test) 
# creating a confusion matrix 
cm = confusion_matrix(y_test, svm_predictions) 
print("Accuracy of SVM is:",accuracy)
print(cm)

Accuracy of SVM is: 0.5405405405405406
[[ 9 11  0]
 [ 3 11  0]
 [ 1  2  0]]


In [61]:
from sklearn.ensemble import RandomForestClassifier
model=RandomForestClassifier(n_estimators=25)
model.fit(x_train,y_train)
preds=model.predict(x_test)
acc=accuracy_score(y_test,preds)
print("The Accuracy of randomforestclassifier is:", acc)

parameters = {
    "n_estimators":[5,10,50,100,250],
    "max_depth":[2,4,8,16,32,None]
}

from sklearn.model_selection import GridSearchCV
cv = GridSearchCV(model,parameters,cv=5)
cv.fit(x_train,y_train)
print("The best accuracy score obtained is:",cv.best_score_," at parameters ", cv.best_params_)


The Accuracy of randomforestclassifier is: 0.5135135135135135
The best accuracy score obtained is: 0.5528735632183908  at parameters  {'max_depth': 32, 'n_estimators': 10}


In [60]:
from sklearn.ensemble import RandomForestClassifier
model=RandomForestClassifier(max_depth=None,n_estimators=5)
model.fit(x_train,y_train)
preds=model.predict(x_test)
acc=accuracy_score(y_test,preds)
print("The Accuracy of randomforestclassifier is:", acc)

The Accuracy of randomforestclassifier is: 0.6216216216216216


In [62]:
from sklearn.naive_bayes import GaussianNB
model=GaussianNB()
model.fit(x_train,y_train)
preds=model.predict(x_test)
acc=accuracy_score(y_test,preds)
print("The Accuracy of Naive Bayes is:", acc)

The Accuracy of Naive Bayes is: 0.2972972972972973


In [11]:
from sklearn.linear_model import LogisticRegression
model=LogisticRegression()
model.fit(x_train,y_train)
preds=model.predict(x_test)
acc=accuracy_score(y_test,preds)
print("The Accuracy of Logistic regression is:", acc)

The Accuracy of Logistic regression is: 0.6486486486486487


c:\Users\sakth\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [13]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler,MinMaxScaler

pipe = make_pipeline(MinMaxScaler(), LogisticRegression())
pipe.fit(x_train, y_train)  # apply scaling on training data
score=pipe.score(x_test, y_test) 
print("The accuracy score of logistic regression is: ", score)

The accuracy score of logistic regression is:  0.6216216216216216


In [14]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler,MinMaxScaler
import pickle

pipe = make_pipeline(MinMaxScaler(), KNeighborsClassifier(n_neighbors=70))
pipe.fit(x_train, y_train)  # apply scaling on training data
score=pipe.score(x_test, y_test)
print(score)
pickle.dump([pipe], open('knn.pkl', 'wb'))

0.7027027027027027


In [ ]:
max_score=0
max_ind=-1
for i in range(1,148):
    pipe = make_pipeline(MinMaxScaler(), KNeighborsClassifier(n_neighbors=i))
    pipe.fit(x_train, y_train)  # apply scaling on training data
    score=pipe.score(x_test, y_test)
    if score>max_score:
        max_score=score
        max_ind=i
print(max_score,max_ind)

Best accuracy is obtained with KNeighboursClassifier model at n-neigbours parameter set to 20.
Accuracy score obtained = 70.27 %

In [15]:
from sklearn.linear_model import Perceptron

model=Perceptron()
model.fit(x_train,y_train)
preds=model.predict(x_test)
acc=accuracy_score(y_test,preds)

print("The accuracy of a Perceptron is:",acc)

The accuracy of a Perceptron is: 0.40540540540540543


In [16]:
from sklearn.neural_network import MLPClassifier
clf=MLPClassifier(random_state=1,max_iter=300)
clf.fit(x_train,y_train)
preds=clf.predict(x_test)
acc=accuracy_score(y_test,preds)
print("The accuracy score obtained by MLP classifier is :",acc)

The accuracy score obtained by MLP classifier is : 0.6216216216216216


In [17]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier

clf=AdaBoostClassifier(
    DecisionTreeClassifier(max_depth=1),n_estimators=100,algorithm="SAMME.R",learning_rate=0.5
    )
clf.fit(x_train,y_train)
preds=clf.predict(x_test)
score=accuracy_score(y_test,preds)

print("The accuracy of decision tree with adaboost is",score)

The accuracy of decision tree with adaboost is 0.4594594594594595
